[TensorExpr] Add classes for memory management in tensor expressions.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33220 [TensorExpr] Add IR Printer.
* #33219 [TensorExpr] Add IR visitor, IR mutator, and IR evaluator.
* #33218 [TensorExpr] Add core classes for representing expressions and statements.
* #33217 [TensorExpr] Add a class for representing data type.
* **#33216 [TensorExpr] Add classes for memory management in tensor expressions.**

All tensor expressions belong to a kernel arena and are freed when the arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

Commit:
[TensorExpr] Add classes for memory management in tensor expressions.

All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Looks good. Also it's been reviewed on another branch.

Commit:
Rebase on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the
arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]

Commit:
Rebase. on "[TensorExpr] Add classes for memory management in tensor expressions."


All tensor expressions belong to a kernel arena and are freed when the arena is destroyed. Until it is destroyed, all expressions stay valid.

Differential Revision: [D19848382](https://our.internmc.facebook.com/intern/diff/D19848382)

[ghstack-poisoned]
